code/game/dna/dna2.dm

```diff
@@ -102,112 +140,152 @@ GLOBAL_LIST_EMPTY(bad_blocks)
 
-	/*SetUIValueRange(DNA_UI_BACC_STYLE,	bodyacc,	GLOB.facial_hair_styles_list.len,	1)*/
-	SetUIValueRange(DNA_UI_HEAD_MARK_STYLE,	head_marks,		GLOB.marking_styles_list.len,		1)
-	SetUIValueRange(DNA_UI_BODY_MARK_STYLE,	body_marks,		GLOB.marking_styles_list.len,		1)
-	SetUIValueRange(DNA_UI_TAIL_MARK_STYLE,	tail_marks,		GLOB.marking_styles_list.len,		1)
+	/*SetDNAValueRange(DNA_UI_BACC_STYLE,	bodyacc,	GLOB.facial_hair_styles_list.len,	DNA_UI,	TRUE)*/
+	SetDNAValueRange(DNA_UI_HEAD_MARK_STYLE,	head_marks,		GLOB.marking_styles_list.len,		DNA_UI,	TRUE)
+	SetDNAValueRange(DNA_UI_BODY_MARK_STYLE,	body_marks,		GLOB.marking_styles_list.len,		DNA_UI,	TRUE)
+	SetDNAValueRange(DNA_UI_TAIL_MARK_STYLE,	tail_marks,		GLOB.marking_styles_list.len,		DNA_UI,	TRUE)
 
 	//Set the Gender
 	switch(character.gender)
 		if(FEMALE)
-			SetUITriState(DNA_UI_GENDER, DNA_GENDER_FEMALE, 1)
+			SetDNATriState(DNA_UI_GENDER, DNA_GENDER_FEMALE, DNA_UI, TRUE)
 		if(MALE)
-			SetUITriState(DNA_UI_GENDER, DNA_GENDER_MALE, 1)
+			SetDNATriState(DNA_UI_GENDER, DNA_GENDER_MALE, DNA_UI, TRUE)
 		if(PLURAL)
-			SetUITriState(DNA_UI_GENDER, DNA_GENDER_PLURAL, 1)
+			SetDNATriState(DNA_UI_GENDER, DNA_GENDER_PLURAL, DNA_UI, TRUE)
 
 
-	UpdateUI()
+	UpdateDNA(DNA_UI)
 
-// Set a DNA UI block's raw value.
-/datum/dna/proc/SetUIValue(block, value, defer = FALSE)
-	if(block <= 0)
+
+/datum/dna/proc/ValidCheck(block, dna_type)
+	if(block <= 0 || dna_type < 0 || dna_type > 2 || dna_type == null)
+		return FALSE
+	else
+		return TRUE
+
+///////////////////////////////////////
+// SET and GET values for DNA blocks
+///////////////////////////////////////
+
+// Set a DNA block's raw value.
+/datum/dna/proc/SetDNAValue(block, value, dna_type, defer = FALSE)
+	if(!ValidCheck(block, dna_type))
 		return
 	ASSERT(value > 0)
 	ASSERT(value <= 4095)
-	UI[block]=value
-	dirtyUI = 1
+	switch(dna_type)
+		if(DNA_UI)
+			UI[block] = value
+			dirtyUI = TRUE
+		if(DNA_SE)
+			SE[block] = value
+			dirtySE = TRUE
+		if(DNA_RP)
+			RP[block] = value
+			dirtyRP = TRUE
 	if(!defer)
-		UpdateUI()
+		UpdateDNA(dna_type)
 
-// Get a DNA UI block's raw value.
-/datum/dna/proc/GetUIValue(block)
-	if(block <= 0)
+// Get a DNA block's raw value.
+/datum/dna/proc/GetDNAValue(block, dna_type)
+	if(!ValidCheck(block, dna_type))
 		return FALSE
-	return UI[block]
-
-// Set a DNA UI block's value, given a value and a max possible value.
-// Used in hair and facial styles (value being the index and maxvalue being the len of the hairstyle list)
-/datum/dna/proc/SetUIValueRange(block, value, maxvalue, defer = FALSE)
-	if(block <= 0)
+	switch(dna_type)
+		if(DNA_UI)
+			return UI[block]
+		if(DNA_SE)
+			return SE[block]
+		if(DNA_RP)
+			return RP[block]
+
+// Set a DNA block's value, given a value and a max possible value
+/datum/dna/proc/SetDNAValueRange(block, value, maxvalue, dna_type, defer = FALSE)
+	if(!ValidCheck(block, dna_type) || !value)
```

Comment:
Does the second conditional (`!value`) prevent DNA blocks from being set to a value of `0` ?

Maybe better to check if `value < 0 || value > maxvalue` ?

Reply:
This proc was originally `SetUIValueRange()`.

I combined it with `SetSEValueRange()` which did not have the same `if(value == 0)` check, so that check only happens if the `dna_type` is `DNA_UI`.

```
/datum/dna/proc/SetSEValueRange(block, value, maxvalue)
	if(block <= 0)
		return
	ASSERT(maxvalue <= 4095)
	var/range = round(4095 / maxvalue)
	if(value)
		SetSEValue(block, value * range - rand(1, range - 1))
```

A-And, hang on . . . hold up . . . 

![image](https://user-images.githubusercontent.com/69871346/107814469-f70b1d00-6d3f-11eb-94fd-9449b08a9792.png)

I do not currently see a use case for using `SetSEValueRange()`, now `SetDNAValueRange(DNA_SE)` and this code supports the fact that using it does not make much sense.  The scope of this PR was mainly to combine the procs which involved changing syntax. This review has brought up some really great questions. At this point, I think it might be best to code supporting setting value ranges for SE (commented out?). But, let's move that checker up some levels.